### PR TITLE
Tenant ID propagation sample

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Create a sample that contains three endpoints
 * Client that sets the tenant ID on the outgoing messages
 * Sales that propagates the tenant ID to the downstream endpoint
 * Billing that shows that the tenant ID propagates

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@mauroservienti I didn't create the proper V7 solutions yet because I would like to first get your feedback on the code & content. Once done, I'll copy the code over to the V7 folders.

**Comment by mauroservienti** [comment]
Faster than light! you rock @SzymonPobiega 😄

**Comment by mauroservienti** [comment]
Code is exactly what I was thinking about. Perfect.
I'll review the `sample.md` document later today.

File: samples/multi-tenant/propagation/sample.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+reviewed: 2019-02-08
+component: Core
+related:
+- nservicebus/pipeline/manipulate-with-behaviors
+---
+
+
+## Introduction
+
+This sample shows how to configure the NServiceBus pipeline to automatically propagate tenant information to the downstream endpoints. The sample assumes that the tenant information is passed as a message header.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
This sample shows how to configure the NServiceBus pipeline to automatically propagate tenant information to downstream endpoints. The sample assumes that the tenant information is passed as a message header.
```

File: samples/multi-tenant/propagation/sample.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+## Code Walk Through
+
+
+### Attaching the tenant information to the message
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
### Attaching tenant information to the messages
```

File: samples/multi-tenant/propagation/sample.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+In most cases the best way to attach tenant information message is by using a custom message header. Following code demonstrates how to set a custom header `tenant_id`.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
In most cases the best way to attach tenant information to messages is by using a custom message header. The following code demonstrates how to set a custom `tenant_id` header.
```

File: samples/multi-tenant/propagation/sample.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: SetTenantId
+
+
+### Creating the Behaviors
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
### Creating Behaviors
```

File: samples/multi-tenant/propagation/sample.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+There are two behaviors required to propagate the tenant information.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
Two behaviors are required to propagate the tenant information.
```

File: samples/multi-tenant/propagation/sample.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+
+#### Retrieving and storing the tenant information
+
+The first behavior is responsible for extracting the tenant information from a message header and placing it in the pipeline execution context bag. This behavior executes as part of the message receive pipeline.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
The first behavior is responsible for extracting the tenant information from the incoming message header, and placing it in the pipeline execution context bag. This behavior executes as part of the message receive pipeline.
```

File: samples/multi-tenant/propagation/sample.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: StoreTenantId
+
+
+#### Propagating the tenant information to the outgoing messages
+
+The second behavior is responsible for attaching the tenant information header(s) to the outgoing messages based on the pipeline context. This behavior executes as part of the message send pipeline.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
The second behavior is responsible for attaching the tenant information header(s) to outgoing messages based on the pipeline context. This behavior executes as part of the message send pipeline.
```

File: samples/multi-tenant/propagation/sample.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+snippet: PropagateTenantId
+
+
+### Registering the behaviors
+
+The following code is needed to register the behavior in the pipeline.
```
Comment (mauroservienti) [inline samples/multi-tenant/propagation/sample.md]:
```suggestion
The following code is needed to register the created behaviors in the pipeline.
```

**Comment by SzymonPobiega** [comment]
@mauroservienti I've applied your suggestions.

**Comment by mauroservienti** [comment]
Thanks @SzymonPobiega 

Is the v7 version the only missing thing then?

**Comment by SzymonPobiega** [comment]
@mauroservienti yeah. I am working on it now. Will ping you when done.

**Comment by SzymonPobiega** [comment]
@mauroservienti I've added the V7 full framework + core

**Pull request merged**